motioncorr/viewers.py

```diff
@@ -24,45 +26,120 @@
-                        'outputMovies']
-
-        if not any(obj.hasAttribute(l) for l in outputLabels):
-            return [self.infoMessage("Output (micrographs or movies) have "
-                                     "not been produced yet.")]
-
-        # Display only the first available output, showing all of them
-        # can be confusing and not useful.
-        # The user can still double-click in the specific output
-        for l in outputLabels:
-            if obj.hasAttribute(l):
-                output = getattr(obj, l)
-                if 'Micrographs' in l:
-                    return [MicrographsView(self.getProject(), output)]
-                else:  # Movies case
-                    return [self.objectView(output, viewParams=viewParamsDef)]
+        elif param == 'doShowMicsDW':
+            if getattr(self.protocol, 'outputMicrographsDoseWeighted', None) is not None:
+                return [MicrographsView(self.getProject(),
+                                    self.protocol.outputMicrographsDoseWeighted)]
+            else:
+                self._errors.append('No output dose-weighted micrographs found!')
 
+        elif param == 'doShowMovies':
+            if getattr(self.protocol, 'outputMovies', None) is not None:
+                labelsDef = 'enabled id _filename _samplingRate '
+                labelsDef += '_acquisition._dosePerFrame _acquisition._doseInitial '
+                viewParamsDef = {showj.MODE: showj.MODE_MD,
+                                 showj.ORDER: labelsDef,
+                                 showj.VISIBLE: labelsDef,
+                                 showj.RENDER: None
+                                 }
+                output = self.protocol.outputMovies
+                return [self.objectView(output, viewParams=viewParamsDef)]
+            else:
+                self._errors.append('No output movies found!')
+
+        elif param == 'doShowFailedMovies':
+            failedList = self.protocol._readFailedList()
+            if not failedList:
+                self._errors.append("No failed movies found!")
+            else:
+                labelsDef = 'enabled id _filename _samplingRate '
+                labelsDef += '_acquisition._dosePerFrame _acquisition._doseInitial '
+                viewParamsDef = {showj.MODE: showj.MODE_MD,
+                                 showj.ORDER: labelsDef,
+                                 showj.VISIBLE: labelsDef,
+                                 showj.RENDER: None
+                                 }
+                sqliteFn = self.protocol._getExtraPath('movies_failed.sqlite')
+                self.createFailedMoviesSqlite(failedList, sqliteFn)
+                return [self.objectView(sqliteFn, viewParams=viewParamsDef)]
+
+    def createFailedMoviesSqlite(self, failedList, path):
+        inputMovies = self.protocol.inputMovies
+        inputSet = [movie.clone() for movie in inputMovies]
```

Comment:
we can discuss if this is a good approach

Reply:
I think you don't need to make this clone. The copyItem approach can be used and it will not incur in any extra performance penalty.

Reply:
so, just create a second set? In principle all I need is to iterate though inputMovies. Using inputMovies directly is not good, right?

Reply:
You are creating a new set (the failed one) in line 131. So, it is fine to use the copyItem function and then set the insert flag or not depending if the movie under inspection is in the failed set or not. Here we are doing two iterations and an unnecessary clone on each element.